File: tachyon/scripts/stand.py
#! /usr/bin/env python
#  Author: David Akhihiero
# Code to make tachyon stand 

import rospy
from std_msgs.msg import Float64
from control_msgs.msg import JointControllerState
from tachyon.msg import JointAnglesErrorIsBelowTolerance
from Leg import Leg


is_at_desired = False
count = 0

def joint_angles_error_callback(msg):
    global is_at_desired
    is_at_desired = msg.is_at_desired

def stand():
    global is_at_desired, count

    base_to_front_left_pub = rospy.Publisher("/tachyon/joint_base_to_front_left_controller/command", Float64, queue_size=10)
    swing_link_to_upper_limb_front_left_pub = rospy.Publisher("/tachyon/joint_swing_link_to_upper_limb_front_left_controller/command", 
    Float64, queue_size=10)
    upper_limb_to_lower_limb_front_left_pub = rospy.Publisher("/tachyon/joint_upper_limb_to_lower_limb_front_left_controller/command", 
    Float64, queue_size=10)

    base_to_front_right_pub = rospy.Publisher("/tachyon/joint_base_to_front_right_controller/command", Float64, queue_size=10)
    swing_link_to_upper_limb_front_right_pub = rospy.Publisher("/tachyon/joint_swing_link_to_upper_limb_front_right_controller/command", 
    Float64, queue_size=10)
    upper_limb_to_lower_limb_front_right_pub = rospy.Publisher("/tachyon/joint_upper_limb_to_lower_limb_front_right_controller/command", 
    Float64, queue_size=10)

    base_to_back_left_pub = rospy.Publisher("/tachyon/joint_base_to_back_left_controller/command", Float64, queue_size=10)
    swing_link_to_upper_limb_back_left_pub = rospy.Publisher("/tachyon/joint_swing_link_to_upper_limb_back_left_controller/command", 
    Float64, queue_size=10)
    upper_limb_to_lower_limb_back_left_pub = rospy.Publisher("/tachyon/joint_upper_limb_to_lower_limb_back_left_controller/command", 
    Float64, queue_size=10)

    base_to_back_right_pub = rospy.Publisher("/tachyon/joint_base_to_back_right_controller/command", Float64, queue_size=10)
    swing_link_to_upper_limb_back_right_pub = rospy.Publisher("/tachyon/joint_swing_link_to_upper_limb_back_right_controller/command", 
    Float64, queue_size=10)
    upper_limb_to_lower_limb_back_right_pub = rospy.Publisher("/tachyon/joint_upper_limb_to_lower_limb_back_right_controller/command", 
    Float64, queue_size=10)
    

    # setting default_height as 0.7m
    default_height = 0.7

    # create Leg objects for each leg, specifying the leg parameters: L1, H1, L2, L3, leg_name and if the leg
    # is on the right or on the left side of tachyon and use the get_joint_angles() method to compute the
    # inverse kinematics

    front_left_leg = Leg (0.13002, 0.08274, 0.39825, 0.33, 'front_left', False)
    angles = front_left_leg.get_joint_angles(0, -default_height, 0.13002)
    theta1, theta2, theta3 = angles[0]

    base_to_front_left_pos = Float64(theta1)
    swing_link_to_upper_limb_front_left_pos = Float64(theta2)
    upper_limb_to_lower_limb_front_left_pos = Float64(theta3)

    front_right_leg = Leg (0.13002, 0.08274, 0.39825, 0.33, 'front_right', True)
    angles = front_right_leg.get_joint_angles(0, -default_height, 0.13002)
    theta1, theta2, theta3 = angles[0]

    base_to_front_right_pos = Float64(theta1)
    swing_link_to_upper_limb_front_right_pos = Float64(theta2)
    upper_limb_to_lower_limb_front_right_pos = Float64(theta3)

    back_left_leg = Leg (0.13002, 0.08274, 0.39825, 0.33, 'back_left', False)
    angles = back_left_leg.get_joint_angles(0, -default_height, 0.13002)
    theta1, theta2, theta3 = angles[0]

    base_to_back_left_pos = Float64(theta1)
    swing_link_to_upper_limb_back_left_pos = Float64(theta2)
    upper_limb_to_lower_limb_back_left_pos = Float64(theta3)

    back_right_leg = Leg (0.13002, 0.08274, 0.39825, 0.33, 'back_right', True)
    angles = back_right_leg.get_joint_angles(0, -default_height, 0.13002)
    theta1, theta2, theta3 = angles[0]

    base_to_back_right_pos = Float64(theta1)
    swing_link_to_upper_limb_back_right_pos = Float64(theta2)
    upper_limb_to_lower_limb_back_right_pos = Float64(theta3)

    rate = rospy.Rate(50)

    standing = False
    while not standing and not rospy.is_shutdown(): 
        base_to_front_left_pub.publish(base_to_front_left_pos)
        swing_link_to_upper_limb_front_left_pub.publish(swing_link_to_upper_limb_front_left_pos)
        upper_limb_to_lower_limb_front_left_pub.publish(upper_limb_to_lower_limb_front_left_pos)

        base_to_front_right_pub.publish(base_to_front_right_pos)
        swing_link_to_upper_limb_front_right_pub.publish(swing_link_to_upper_limb_front_right_pos)
        upper_limb_to_lower_limb_front_right_pub.publish(upper_limb_to_lower_limb_front_right_pos)

        base_to_back_left_pub.publish(base_to_back_left_pos)
        swing_link_to_upper_limb_back_left_pub.publish(swing_link_to_upper_limb_back_left_pos)
        upper_limb_to_lower_limb_back_left_pub.publish(upper_limb_to_lower_limb_back_left_pos)

        base_to_back_right_pub.publish(base_to_back_right_pos)
        swing_link_to_upper_limb_back_right_pub.publish(swing_link_to_upper_limb_back_right_pos)
        upper_limb_to_lower_limb_back_right_pub.publish(upper_limb_to_lower_limb_back_right_pos)
            
        rate.sleep()
        standing = is_at_desired
    
    if count < 3:
        count += 1
        stand()

        


if __name__ == "__main__":
    rospy.init_node("stand_node", anonymous=True)
    rospy.wait_for_message("/tachyon/joint_upper_limb_to_lower_limb_back_right_controller/state", JointControllerState)
    rospy.Subscriber("joint_angles_error_publisher", JointAnglesErrorIsBelowTolerance, joint_angles_error_callback)
    stand()

File: tachyon/src/stand.cpp
// Author: David Akhihiero

// Cpp Code to make tachyon stand 

#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <control_msgs/JointControllerState.h>
#include <tachyon/JointAnglesErrorIsBelowTolerance.h>
#include "Leg.h"

bool isAtDesired = false;
int count = 0;
float L1 = 0.1, H1 = 0.0, L2 = 0.4, L3 = 0.4;

void jointAnglesErrorCallback(const tachyon::JointAnglesErrorIsBelowTolerance::ConstPtr &msg)
{
    isAtDesired = msg->is_at_desired;
}

void stand()
{
    ros::NodeHandle nh;

    ros::Publisher baseToFrontLeftPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_base_to_front_left_controller/command", 10);
    ros::Publisher swingLinkToUpperLimbFrontLeftPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_swing_link_to_upper_limb_front_left_controller/command", 10);
    ros::Publisher upperLimbToLowerLimbFrontLeftPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_upper_limb_to_lower_limb_front_left_controller/command", 10);

    ros::Publisher baseToFrontRightPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_base_to_front_right_controller/command", 10);
    ros::Publisher swingLinkToUpperLimbFrontRightPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_swing_link_to_upper_limb_front_right_controller/command", 10);
    ros::Publisher upperLimbToLowerLimbFrontRightPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_upper_limb_to_lower_limb_front_right_controller/command", 10);

    ros::Publisher baseToBackLeftPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_base_to_back_left_controller/command", 10);
    ros::Publisher swingLinkToUpperLimbBackLeftPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_swing_link_to_upper_limb_back_left_controller/command", 10);
    ros::Publisher upperLimbToLowerLimbBackLeftPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_upper_limb_to_lower_limb_back_left_controller/command", 10);

    ros::Publisher baseToBackRightPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_base_to_back_right_controller/command", 10);
    ros::Publisher swingLinkToUpperLimbBackRightPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_swing_link_to_upper_limb_back_right_controller/command", 10);
    ros::Publisher upperLimbToLowerLimbBackRightPub = nh.advertise<std_msgs::Float64>("/tachyon/joint_upper_limb_to_lower_limb_back_right_controller/command", 10);

    // setting default_height as 0.6m
    float defaultHeight = 0.6;

    // create Leg objects for each leg, specifying the leg parameters: L1, H1, L2, L3, leg_name and if the leg
    // is on the right or on the left side of tachyon and use the get_joint_angles() method to compute the
    // inverse kinematics

    Leg frontLeftLeg(L1, H1, L2, L3, "front_left", false);
    std::array<float, 3> angles = frontLeftLeg.getJointAngles(0, -defaultHeight, L1);
    float theta1 = angles[0], theta2 = angles[1], theta3 = angles[2];

    std_msgs::Float64 baseToFrontLeftPos;
    std_msgs::Float64 swingLinkToUpperLimbFrontLeftPos;
    std_msgs::Float64 upperLimbToLowerLimbFrontLeftPos;
    baseToFrontLeftPos.data = theta1;
    swingLinkToUpperLimbFrontLeftPos.data = theta2;
    upperLimbToLowerLimbFrontLeftPos.data = theta3;
    


    Leg frontRightLeg(L1, H1, L2, L3, "front_right", true);
    angles = frontRightLeg.getJointAngles(0, -defaultHeight, L1);
    theta1 = angles[0], theta2 = angles[1], theta3 = angles[2];

    std_msgs::Float64 baseToFrontRightPos;
    std_msgs::Float64 swingLinkToUpperLimbFrontRightPos;
    std_msgs::Float64 upperLimbToLowerLimbFrontRightPos;
    baseToFrontRightPos.data = theta1;
    swingLinkToUpperLimbFrontRightPos.data = theta2;
    upperLimbToLowerLimbFrontRightPos.data = theta3;
    

    Leg backLeftLeg(L1, H1, L2, L3, "back_left", false);
    angles = backLeftLeg.getJointAngles(0, -defaultHeight, L1);
    theta1 = angles[0], theta2 = angles[1], theta3 = angles[2];

    std_msgs::Float64 baseToBackLeftPos;
    std_msgs::Float64 swingLinkToUpperLimbBackLeftPos;
    std_msgs::Float64 upperLimbToLowerLimbBackLeftPos;
    baseToBackLeftPos.data = theta1;
    swingLinkToUpperLimbBackLeftPos.data = theta2;
    upperLimbToLowerLimbBackLeftPos.data = theta3;
    


    Leg backRightLeg(L1, H1, L2, L3, "back_right", true);
    angles = backRightLeg.getJointAngles(0, -defaultHeight, L1);
    theta1 = angles[0], theta2 = angles[1], theta3 = angles[2];

    std_msgs::Float64 baseToBackRightPos;
    std_msgs::Float64 swingLinkToUpperLimbBackRightPos;
    std_msgs::Float64 upperLimbToLowerLimbBackRightPos;
    baseToBackRightPos.data = theta1;
    swingLinkToUpperLimbBackRightPos.data = theta2;
    upperLimbToLowerLimbBackRightPos.data = theta3;

    ros::Rate rate(50);

    bool standing = false;

    while (!standing && ros::ok())
    {
        baseToFrontLeftPub.publish(baseToFrontLeftPos);
        swingLinkToUpperLimbFrontLeftPub.publish(swingLinkToUpperLimbFrontLeftPos);
        upperLimbToLowerLimbFrontLeftPub.publish(upperLimbToLowerLimbFrontLeftPos);

        baseToFrontRightPub.publish(baseToFrontRightPos);
        swingLinkToUpperLimbFrontRightPub.publish(swingLinkToUpperLimbFrontRightPos);
        upperLimbToLowerLimbFrontRightPub.publish(upperLimbToLowerLimbFrontRightPos);

        baseToBackLeftPub.publish(baseToBackLeftPos);
        swingLinkToUpperLimbBackLeftPub.publish(swingLinkToUpperLimbBackLeftPos);
        upperLimbToLowerLimbBackLeftPub.publish(upperLimbToLowerLimbBackLeftPos);

        baseToBackRightPub.publish(baseToBackRightPos);
        swingLinkToUpperLimbBackRightPub.publish(swingLinkToUpperLimbBackRightPos);
        upperLimbToLowerLimbBackRightPub.publish(upperLimbToLowerLimbBackRightPos);
        rate.sleep();
        ros::spinOnce();

        standing = isAtDesired;

    }

    if (count < 3)
    {
        count++;
        stand();
    }
}


int main(int argc, char** argv)
{
    ros::init(argc, argv, "stand_node_cpp");
    ros::NodeHandle nh;

    const std::string msg = "/tachyon/joint_upper_limb_to_lower_limb_back_right_controller/state";

    ros::topic::waitForMessage<control_msgs::JointControllerState>(msg);

    ros::Subscriber sub = nh.subscribe("joint_angles_error_publisher", 10, jointAnglesErrorCallback);

    ros::spinOnce();

    stand();

}

